File: app/src/main/java/com/bidsaarthi/ai/data/BasicAnalysis.kt
package com.bidsaarthi.ai.data

import com.bidsaarthi.ai.model.Tender
import org.json.JSONArray
import org.json.JSONObject

/** Free offline checks, intentionally never presented as generative AI. */
fun basicAnalysis(t:Tender,profile:JSONObject,reason:String=""):JSONObject {
 val categories=profile.optJSONArray("categories") ?: JSONArray()
 val hay="${t.title} ${t.department} ${t.evidence}"
 val hits=(0 until categories.length()).map { categories.optString(it) }.filter { it.isNotBlank() && hay.contains(it,true) }
 val reasons=JSONArray().put("Basic rule-based check; no AI model was used.")
 reasons.put(if(hits.isEmpty()) "No business keywords matched the available listing." else "Matching keywords: ${hits.distinct().joinToString()}")
 if(reason.isNotBlank()) reasons.put(reason)
 val risks=JSONArray().put("Verify the full tender, fees, exemptions and amendments on the official portal.")
 val remaining=deadlineMillis(t.deadline)?.minus(System.currentTimeMillis())
 if(remaining!=null && remaining<0) risks.put("The listed deadline has passed.")
 else if(remaining!=null && remaining<=172800000L) risks.put("The listed deadline is within 48 hours.")
 return JSONObject().put("analysis_mode","RULE_BASED").put("provider","Offline basic checks").put("eligibility","UNKNOWN")
  .put("summary",t.title).put("confidence",0).put("eligibility_reasons",reasons).put("risks",risks)
  .put("missing_documents",JSONArray().put("Full eligibility clauses and document requirements need verification."))
  .put("evidence",JSONObject().apply { if(t.evidence.isNotBlank()) put("listing",t.evidence.take(600)) })
}

File: backend/app/basic_analysis.py
"""Free deterministic checks. These are explicitly not generative AI."""
from datetime import datetime
from zoneinfo import ZoneInfo
from .models import Analysis, BusinessDNA


def basic_analysis(tender, business=None, reason=''):
    business = business or BusinessDNA()
    source = tender.evidence.get('listing', '')
    text = (tender.title + ' ' + tender.department + ' ' + source).lower()
    hits = [k for k in business.categories + business.keywords if k.strip() and k.lower() in text]
    reasons = ['Basic rule-based check; no AI model was used.']
    if hits:
        reasons.append('Matching business keywords: ' + ', '.join(dict.fromkeys(hits)))
    else:
        reasons.append('No matching business keywords found in the available listing.')
    if reason:
        reasons.append(reason)
    risks = ['Verify the full official tender, amendments, fees and exemptions before bidding.']
    for fmt in ('%d-%b-%Y %I:%M %p', '%d-%b-%Y %H:%M'):
        try:
            closes = datetime.strptime(tender.closes_at or '', fmt).replace(tzinfo=ZoneInfo('Asia/Kolkata'))
            hours = (closes - datetime.now(ZoneInfo('Asia/Kolkata'))).total_seconds() / 3600
            if hours < 0: risks.append('The listed submission deadline has passed.')
            elif hours <= 48: risks.append('The listed submission deadline is within 48 hours.')
            break
        except ValueError:
            continue
    return Analysis(tender_id=tender.id, analysis_mode='RULE_BASED', provider='on-device-compatible rules',
                    eligibility='UNKNOWN', summary=tender.title,
                    eligibility_reasons=reasons, risks=risks,
                    missing_documents=['Full eligibility clauses and document requirements need verification.'],
                    evidence={'listing': source[:600]} if source else {}, confidence=0)
